Doc/whatsnew/3.11.rst

```diff
@@ -493,6 +494,222 @@ Optimizations
+
+(Contributed by Eric Snow, Guido van Rossum and Kumar Aditya in numerous issues.)
+
+
+Faster Runtime
+--------------
+
+Cheaper, lazy Python frames
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Python frames are created whenever Python calls a Python function. This frame
+holds execution information. The following are new frame optimizations:
+
+- Streamlined the frame creation process.
+- Avoided memory allocation by generously re-using frame space on the C stack.
+- Streamlined the internal frame object to only contain essential information.
+  Frames previously held extra debugging and memory management information.
+  Old-style frames are now created only when required by debuggers. For most
+  user code, no frames are created at all.
+
+Nearly all Python functions calls have sped up significantly.
+This resulted in a 3-7% improvement in pyperformance.
+(Contributed by Mark Shannon in :issue:`44590`.)
+
+.. _inline-calls:
+
+Inlined Python function calls
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+During a Python function call, Python will call a C function to interpret that
+function's code, similar to how a user would call ``eval`` to run arbitrary
+code.
+
+In 3.11, when Python detects code calling another Python function,
+it sets up a new frame, and "jumps" to the new code inside the new frame. This
+avoids calling the C interpreting function altogether.
+
+Python function calls now consume almost no C stack space. This speeds up Python
+to Python function calls. In simple recursive functions like fibonacci or
+factorial, a 1.7x speedup was observed. This also means recursive functions
+can recurse significantly deeper, assuming the recursion limit and memory limit
+is not exceeded. This resulted in a 1-3% improvement in pyperformance.
+(Contributed by Pablo Galindo and Mark Shannon in :issue:`45256`.)
+
+PEP 659: Specializing Adaptive Interpreter
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+:pep:`659` is one of the key parts of the faster CPython project. The general
+idea is that while Python is a dynamic language, most code have regions where
+objects and types rarely change. This concept is known as *type stability*.
+
+At runtime, Python will try to look for common patterns and type stability
+in the executing code. Python will then replace the current operation with a
+more specialized one. This specialized operation use fast paths available only
+to those use cases/types and will generally outperform their generic
+counterparts. This also brings in another concept called *inline caching*, where
+Python caches the results of expensive operations directly in the bytecode.
+
+The specializer will also combine certain common instruction pairs into one
+super instruction. This reduces the overhead during execution.
+
+Since this exta information requires more memory, Python will only specialize
+when it sees code that is "hot" (executed multiple times). It can also
+de-specialize when code is too dynamic. This is attempted every so often, and
```

Comment:
Not only when too dynamic, but when the use changes. The idea is that the code adapts to the new circumstances.